### ISSUE TITLE
Restructuration globale du code

### ISSUE THREAD

**Issue by FlxPo** [mobility-team/mobility#130]
Je crée une nouvelle issue globale pour suivre la restructuration du code qui devrait permettre de répondre aux issues actuellement ouvertes. Je suis reparti de la branche carpool et ait ajouté toutes les modifications des autres branches pour avoir la version la plus à jour du code.

La modification la plus importante à ce stade est la création de groupes de classes par mode de transport:
- Une classe de type ModeParameters, pour stocker toutes les hypothèses nécessaires.
- Une classe de type PathTravelCosts ou PublicTransportTravelCosts, pour faire les calculs de distances, temps, coûts généralisé.
- Une classe chapeau de type TransportMode, qui stocke une instance de ModeParameters et TravelCosts.

Cela permet de bien séparer les hypothèses entre modes, et d'avoir un endroit unique où fixer des hypothèses par défaut. Cela permet aussi de créer facilement des modes dérivés, comme le covoiturage (basé sur le mode voiture mais avec une classe CarpoolTravelCosts spécifique), ou un mode "multimodal" qui contient les coûts de tous les modes : 

```
class MultiModalMode(TransportMode):
    
    def __init__(
        self,
        transport_zones: TransportZones,
        modes: Optional[List[TransportMode]] = None
    ):
        
        if modes is None:
            
            car_mode = CarMode(transport_zones)
            
            modes = [
                WalkMode(transport_zones),
                BicycleMode(transport_zones),
                PublicTransportMode(transport_zones),
                car_mode,
                CarpoolMode(car_mode, CarpoolParameters(number_persons=2)),
                CarpoolMode(car_mode, CarpoolParameters(number_persons=3)),
                CarpoolMode(car_mode, CarpoolParameters(number_persons=4))
            ]
        
        travel_costs = MultiModalTravelCosts(transport_zones, modes)
        super().__init__(travel_costs, ModeParameters("multimodal"))
```

Cela devrait aussi faciliter l'application de "leviers", puisqu'il suffira de créer autant de versions de ModeParameters que l'on a de variantes du modèle.

Autre petits changements :
- Début de regroupement des fichiers python/R par "fonction".
- Ajout d'un dossier "experiments" pour stocker des scripts en cours de développement, pas encore terminés ni intégrés à l'architecture générale de Mobility.
- Amélioration du nom de certaines classes : TravelCosts -> PathTravelCosts (pour bien faire la différence avec PublicTransportTravelCosts).

@Mind-the-Cap

**Comment by Mind-the-Cap** [comment]
Pour l'instant je tente de faire tourner `small_run` de Mobility Grand Genève, j'ai un premier bug pas très clair qui fait que `local_admin_unit_id_to` devient identique à `local_admin_unit_id_from` dans tous les cas, et qui semble venir des travel_costs de covoiturage qui sont ensuite fusionnés grâce à `aggregate_travel_costs`

**Comment by Mind-the-Cap** [comment]
Oui, déjà tenté ! Je vais implémenter la colonne, elle me semble manquer et ça pourrait résoudre le problème.

**Comment by Mind-the-Cap** [comment]
Autre solution car il y avait besoin des admin_ids pour certaines fonctionnalités du covoiturage : corriger le merge ([commit 1](https://github.com/mobility-team/mobility/commit/5e655b2bff085edb98b417755bc4041820d8a84e), [commit 2](https://github.com/mobility-team/mobility/commit/263e5912afe2d3149396e4ac045c7a4d207564ac))

**Comment by Mind-the-Cap** [comment]
Toutes les instances de ModeParameters sont bien destinées à être appelées dans le `run`, n'est-ce pas ?

```
car_params = mobility.CarParameters(
    cost_of_time_c0_short = 35.82,
    cost_of_time_c0 = 24.86,
    cost_of_time_c1 = 0.5)
```
par exemple.

Dans ce cas, est-on d'accord qu'il faut les rajouter dans tout un tas de fichiers __init__ ? Est-ce qu'on souhaite les appeler comme `mobility.CarParameters` ou `mobility.transport_modes.CarParameters` ?

**Comment by FlxPo** [comment]
Potentiellement oui, même si l'utilisateur doit pouvoir aussi laisser les valeurs par défaut et ne rien spécifier du tout.

Et oui je serai plutot pour exposer tout ce qui est utile à l'utilisateur au plus haut niveau (mobility.CarParameters), ce qui me semble le plus courant et le plus simple ?
